### PR TITLE
[#1015] port (window-top-line)

### PR DESCRIPTION
**Pull request by krgn**

### CODE REVIEW

File: rust_src/src/windows.rs [hunk 1]
```diff
@@ -1134,4 +1134,16 @@ pub fn select_window_lisp(window: LispObject, norecord: LispObject) -> LispObjec
     unsafe { select_window(window, norecord, false) }
 }
 
+/// Return top line of window WINDOW.
+/// This is the distance, in lines, between the top of WINDOW and the top
+/// of the frame's window area.  For instance, the return value is 0 if
+/// there is no window above WINDOW.
+///
+/// WINDOW must be a valid window and defaults to the selected one.
+#[lisp_fn(min = "0")]
```
[the remark this answers is not in the document]
Comment (krgn) [reply rust_src/src/windows.rs]: I saw this type used this way, so is the way I did it incorrect?

[the remark this answers is not in the document]
Comment (shaleh) [reply rust_src/src/windows.rs]: The C function `decode_valid_window` is not looking for a live window.
This is the cheat sheet:

decode_valid_window -> LispWindowValidOrSelected
decode_live_window -> LispWindowLiveOrSelected
decode_any_window -> LispWindowOrSelected

[the remark this answers is not in the document]
Comment (krgn) [reply rust_src/src/windows.rs]: I see, thank you for the explanation. I'm currently writing tests and will incorporate this.

[the remark this answers is not in the document]
Comment (shaleh) [reply rust_src/src/windows.rs]: :+1:

File: test/src/window-tests.el [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+;; but WITHOUT ANY WARRANTY; without even the implied warranty of
+;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+;; GNU General Public License for more details.
+
+;; You should have received a copy of the GNU General Public License
+;; along with GNU Emacs.  If not, see <https://www.gnu.org/licenses/>.
+
+(require 'ert)
+
+(ert-deftest windown-tests--window-top-line ()
```
Comment (krgn) [inline test/src/window-tests.el]: @shaleh does this test look OK to you?

**Comment by krgn** [comment]
That was a shoddy copy/paste job there, agreed :roll_eyes: 

Broke up the tests as well.

**Comment by krgn** [comment]
Should be OK now, what do y'all think?

**Pull request merged**